### PR TITLE
NEO-2435: Global Custom Filter improvements

### PR DESCRIPTION
**Pull request by enrique-prado**
[Link to updated story in Deploy Preview](https://deploy-preview-560--neo-react-library-storybook.netlify.app/?path=/story/components-table--custom-basic-table-filter-drawer)

**Before tagging the team for a review, I have done the following:**

- [x] run `yarn all` locally: ensures that all tests pass, formatting is done, types pass, and builds pass
- [x] reviewed my code changes
- [x] updated the Deploy Preview link at the top of this PR (or remove it if not applicable)
- [x] added any important information to this PR (description, images, GIFs, ect.)
- [ ] done an accessibility check on my work (tested with Chrome's `axe Dev Tools`, Mac's VoiceOver, etc.)
- [x] tagged `@avaya-dux/dux-devs`

This PR adds the ability to Apply multiple column filters while providing a custom UI for specifying those filters.

- This is mainly accomplished by exposing the setAllFilters() Table instance api.
- The story shows how to provide custom UI and apply filters.
- To clear previously applied filters, click the x clear button in the TextInput and apply.
image

## Summary by CodeRabbit

- **New Features**
  - Introduced multiple new story exports for the Table component, including filtering, row selection, and pagination functionalities.
  - Added advanced filtering capabilities with a new filter drawer and the ability to pre-select rows.

- **Functionality Enhancements**
  - Enhanced state management for filters and row selections, improving interactivity and user experience.
  - Improved filtering capabilities with the addition of a new `allFilters` prop for dynamic filter application.

- **Bug Fixes**
  - Improved handling of row deletions and data synchronization to ensure accurate table updates.

- **Documentation**
  - Updated type definitions to include new properties for enhanced flexibility in the Table component.

### CODE REVIEW

File: src/components/Table/Table.stories.tsx [hunk 1]
```diff
@@ -863,22 +863,57 @@
 	const closeFilterDrawer = () => {
 		setOpenColumnsFilterDrawer(false);
 	};
+
+	type ColumnFilter = { id: string | null, value: string | number | null };
+	const [filtersToApply, setFiltersToApply] = useState<ColumnFilter[]>([]);
```
[the remark this answers is not in the document]
Comment (enrique-prado) [reply src/components/Table/Table.stories.tsx:867]: fixed.

File: src/components/Table/Table.stories.tsx [hunk 2]
```diff
@@ -863,22 +863,56 @@ export const CustomBasicTableFilterDrawer = () => {
...
+				value: nameFilter || null,
+			},
+			{
+				id: "other",
+				value: otherFilter || null,
+			},
+		];
+		setFiltersToApply(filters);
+		closeFilterDrawer();
+	};
```
[the remark this answers is not in the document]
Comment (enrique-prado) [reply src/components/Table/Table.stories.tsx:885]: mostly irrelevant suggestions.

File: src/components/Table/Table.stories.tsx [hunk 3]
```diff
@@ -863,22 +863,57 @@ export const CustomBasicTableFilterDrawer = () => {
...
-				<p>Custom content goes here</p>
-				<br />
-				<p> 'Cancel' and 'Apply' buttons are built-in</p>
+				<TextInput
+					clearable
+					label="Name"
+					type="text"
+					onChange={(e) => setNameFilter(e.currentTarget.value)}
+				/>
+
```
[the remark this answers is not in the document]
Comment (enrique-prado) [reply src/components/Table/Table.stories.tsx:909]: functionality works as expected.

File: src/components/Table/Table.tsx [hunk 4]
```diff
@@ -196,14 +197,23 @@ export const Table = <T extends Record<string, any>>({
...
+		setAllFilters,
 	} = instance;
 
+	// biome-ignore lint/correctness/useExhaustiveDependencies: setAllFilters and filters don't need to be checked for perf reasons.
+	useEffect(() => {
+		logger.info(allFilters, filters);
+
+		// @ts-expect-error
+		setAllFilters(allFilters);
+	}, [allFilters]);
```
[the remark this answers is not in the document]
Comment (enrique-prado) [reply src/components/Table/Table.tsx]: removed.

File: src/components/Table/types/TableTypes.ts [hunk 5]
```diff
@@ -76,30 +76,31 @@ export type RowHeight = "compact" | "medium" | "large";
...
+		manualPagination?: boolean;
+		manualRowCount?: number;
+		manualColumnFilters?: boolean;
+		manualSortBy?: boolean;
+		onManualSortBy?: (columnId: string, sortType: SortType) => void;
+		showPagination?: boolean;
+		pushPaginationDown?: boolean;
+		draggableRows?: boolean;
+		handlePageChange?: (pageIndex: number, pageSize: number) => void;
+		allFilters?: any[];
```
[the remark this answers is not in the document]
Comment (enrique-prado) [reply src/components/Table/types/TableTypes.ts]: fixed

**Review by joe-s-avaya — APPROVED** [review]
consider adding unit tests, but otherwise this all looks good

**Pull request merged**